projects/lights_control/test/test_lights_gpio.c

```diff
@@ -0,0 +1,127 @@
+#include "event_queue.h"
+#include "gpio.h"
+#include "misc.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+#include "lights_events.h"
+#include "lights_gpio.h"
+#include "lights_gpio_config.h"
+
+// Creating a mock lights_gpio config.
+
+// Mock events.
+typedef enum {
+  TEST_LIGHTS_GPIO_MOCK_EVENT_1 = 0,          //
+  TEST_LIGHTS_GPIO_MOCK_EVENT_2,              //
+  TEST_LIGHTS_GPIO_MOCK_EVENT_3,              //
+  TEST_LIGHTS_GPIO_MOCK_EVENT_4_UNSUPPORTED,  //
+  NUM_TEST_LIGHTS_GPIO_MOCK_EVENTS            //
+} TestLightsGPIOMockEvent;
+
+// Mock peripheral definitions.
+typedef enum {
+  TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_1 = 0,  //
+  TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_2,      //
+  TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_3,      //
+  TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_4,      //
+  TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_5,      //
+  NUM_TEST_LIGHTS_GPIO_MOCK_PERIPHERALS    //
+} TestLightsGPIOMockPeripheral;
+
+// Mock peripheral gpio addresses.
+static const GPIOAddress s_mock_addresses[] = {
+  [TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_1] = { .port = GPIO_PORT_A, .pin = 0 },  //
+  [TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_2] = { .port = GPIO_PORT_B, .pin = 0 },  //
+  [TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_3] = { .port = GPIO_PORT_A, .pin = 4 },  //
+  [TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_4] = { .port = GPIO_PORT_B, .pin = 1 },  //
+  [TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_5] = { .port = GPIO_PORT_A, .pin = 3 },  //
+};
+
+// Array of event-to-peripheral-set mappings.
+static const LightsGPIOEventMapping s_mock_event_mappings[] = {
+  { .event_id = TEST_LIGHTS_GPIO_MOCK_EVENT_1,
+    .peripheral_mapping = LIGHTS_GPIO_PERIPHERAL_BIT(TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_1) |
+                      LIGHTS_GPIO_PERIPHERAL_BIT(TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_3) },                            //
+  { .event_id = TEST_LIGHTS_GPIO_MOCK_EVENT_2,
+    .peripheral_mapping = LIGHTS_GPIO_PERIPHERAL_BIT(TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_2) },  //
+  { .event_id = TEST_LIGHTS_GPIO_MOCK_EVENT_3,
+    .peripheral_mapping = LIGHTS_GPIO_PERIPHERAL_BIT(TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_4) |
+                      LIGHTS_GPIO_PERIPHERAL_BIT(TEST_LIGHTS_GPIO_MOCK_PERIPHERAL_5) }  //
+};
+
+static const LightsGPIO s_mock_config = {
+  .peripherals = s_mock_addresses,                              //
+  .num_peripherals = SIZEOF_ARRAY(s_mock_addresses),            //
+  .event_mappings = s_mock_event_mappings,                    //
+  .num_event_mappings = SIZEOF_ARRAY(s_mock_event_mappings),  //
+};
+
+// Asserts that the gpio state for all peripherals is initialized to high.
+static void prv_gpio_initialized_high(const GPIOAddress *addrs, uint8_t size) {
+  GPIOState state;
+  for (uint8_t i = 0; i < size; i++) {
+    TEST_ASSERT_OK(gpio_get_state(&addrs[i], &state));
+    TEST_ASSERT_EQUAL(state, GPIO_STATE_HIGH);
+  }
+}
+
+// Searches event-mappings for a mapping matching event e.
+static StatusCode prv_search_mappings_table(LightsGPIO *conf, const Event *e, LightsGPIOPeripheralMapping *mapping) {
```

Comment:
These look like you are reusing private functions in a test that invokes a public API. That a bit weird. Usually you shouldn't have to test internal implementation details in this way. Ideally you should only test the public API. If you have concerns about the private API you should expose it via a test_only method. I'm not a big fan of that though so I'd rather you only test the public API.